src/lib/structures/expr.mli

```diff
@@ -309,19 +309,25 @@ val make_triggers:
 (** [make_triggers f binders decl menv] generate multi-triggers for the
     formula [f] and binders [binders].
 
-    For axioms, we try to produce multi-triggers in the following order
-    (if we succeed to produce at least one multi-trigger, we don't try other
-    strategies):
-    1. In greedy mode:
-    - Mono-triggers without variables;
-    - Mono-triggers with variables;
-    - Multi-triggers without variables;
-    - Multi-triggers with variables.
-      2. Otherwise:
-    - Mono and multi-triggers without variables;
-    - Mono and multi-triggers with variables.
-
-    Mono-trigger with `Sy.Plus` or `Sy.Minus` top symbols are ignored
+    We can escape unbounded variables in patterns of multi-triggers, that
+    is replaced them by the underscore term. For instance, if [binders] is
+    the set [{x, y}] and [g(x, y, z)] is a pattern where [{(x, y, z)}] are
+    free term variables, we can replace [z] by [_].
+
+    Our strategy for multi-trigger generations depends on the matching
+    environment [menv] as follows:
+
+    If [menv.greedy] is [false], we try to generate in order:
+    - Mono-triggers;
+    - Multi-triggers with escaping variables.
```

Comment:
Should that be `without`? Does "with escaping variables" means "we allow underscore terms" and "without escaping variables" means "we do not use the underscore term"?

Reply:
`with escaping variables` means we have replaced variables not bound at the top level of this formula by the underscore variable. Here, `escaping` has to be understand as in string context: in the string "aaa\\"aaa", the inner " is escaped.

The name is really confusing because we escape variables that escape the top level binders...

It's `with` and not `without`. Actually we compute multi-triggers without escaping variables but these triggers are threw away:
```ocaml
        let mono, multi =
          make_triggers mconf vterm vtype trs ~escaped_vars:false
        in
        triggers_of_list @@
        match mono with
        | _::_ -> mono (* try to take nb_triggers ? *)
        | [] ->
          let mono', multi' =
            make_triggers mconf vterm vtype f_trs ~escaped_vars:true
          in
          if mono' != [] then mono'
          else if multi != [] then multi
          else multi'
```
If `mono` isn't empty, we ignore the multi-trigger without escaping variables.
If `mono` is empty, we regenerate everything with escaping variables.

In the greedy mode, `mono` and `multi` lists are merged in `make_triggers`.... (terrible design)

Reply:
> with escaping variables means we have replaced variables not bound at the top level of this formula by the underscore variable.

The part I do not understand is if this is exclusive -- are "multi triggers without escaping variables" a subset of "multi triggers with escaping variables" or are they two distinct sets of multi-triggers.

> Here, escaping has to be understand as in string context: in the string "aaa\"aaa", the inner " is escaped.

This confuses me even more. I thought that "escaping variable" referred to a variable escaping its scope (e.g. in `forall x, P(x) -> forall y, Q(x, y)` we must move `Q(x, y)` to the scope of `forall x`, which causes `y` to escape its `forall y` scope). I don't understand the link with character escaping in a string context.

Reply:
> The part I do not understand is if this is exclusive -- are "multi triggers without escaping variables" a subset of "multi triggers with escaping variables" or are they two distinct sets of multi-triggers.

They are distinct.

> This confuses me even more. I thought that "escaping variable" referred to a variable escaping its scope (e.g. in `forall x, P(x) -> forall y, Q(x, y)` we must move `Q(x, y)` to the scope of `forall x`, which causes `y` to escape its `forall y` scope). I don't understand the link with character escaping in a string context.

Uhm, I guess the flag `escaped_var` in `Expr` means `without_escaped_var`? I'm also confused

Reply:
Is it better with the last commit?

Reply:
Yes that looks better!

(Are you sure they are distinct? The only difference seems to be this line `if escaped_vars  then List.rev_map (underscore bv) l else l in` which I think have the same multi-triggers with and without the `escaped_vars` flag for quantifiers with no escapes?

Reply:
It's the only difference but `l` is the list of multi-triggers and we apply substitutions on them. We don't keep the original multi-triggers. I guess I don't understand your question.

Reply:
What I was asking is: are the triggers returned by `make_triggers ~escaped_vars:true` and `make_triggers ~escaped_vars:false` distinct (i.e. no trigger can be in both lists) or not. I understand this is not the case, because triggers with no escaped variables (i.e. only toplevel variables) appear in both lists.

Reply:
Ok... you mean: are these sets *disjoint*? no but these sets are distinct in general.

Reply:
Oh my bad! Yes I meant "disjoint", not "distinct", in spite of making the same mistake twice (: